### ISSUE TITLE
toPromise() Marked as deprecated

### ISSUE THREAD

**Issue by Izhaki** [nestjs/axios#174]
### Did you read the migration guide?

- [X] I have read the whole migration guide

### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Versions

Nest 7 > 8

### Describe the regression

# No Changes Promise

We have recently bumped NestJS from 7 to 8.

The [migration guide](https://docs.nestjs.com/migration-guide#http-module) states:

> The `HttpModule` exported from the `@nestjs/common` package has been deprecated and will be removed in the next major release. Instead, please use the `@nestjs/axios` package (otherwise, there are no API differences).

Note:

> (otherwise, there are no API differences)

# Deprecation Warning

However, with `@nestjs/axios` we get a deprecation warning for `toPromise()`:

![image](https://user-images.githubusercontent.com/880132/136184268-bb6ead71-de22-4a04-a5d6-7f6de841663e.png)

**Note that this deprecation is due to [changes to the RxJS API](https://rxjs.dev/deprecations/to-promise) - not NestJS.**

# Issues

## Quite an API change

Our code is promise based.

With `toPromise()` being deprecated, we are left with an `Observable` - this is quite a big API change!

## Conceptually Dubious

Although I'm sure there's a good reason for this, I'm not sure why methods like `get` or `post` will have an `Observable` API - these operations seem to me, for the consumer at least, more of a `success` or `failure` case, rather than a stream case.

## Axios is a Promise Based Framework

From their [About](https://github.com/axios/axios):

> Promise based HTTP client for the browser and node.js

It's odd to have an `Observable` API from `@nestjs/axios` where `axios` is promise based.

## Not Clear how to Act

Other than having to change a lot of instances of `toPromise()`, it is not clear what is the recommended solution here.

As the [RxJS docs state](https://rxjs.dev/deprecations/to-promise) you need to choose between `lastValueFrom` or `firstValueFrom`. But which one? The model is broken here, we just want to send a request and await the response - no streams involved.

Alternatively, some [SO answers](https://stackoverflow.com/a/60908801/1179377) suggest the following (which doesn't seem to work, by the way):

```js
    return this.http.get(this.DATA_URL).pipe(
      map((axiosResponse: AxiosResponse) => {
        return axiosResponse.data;
      })
    );
```

# Conclusion

To summarise:

- Migration guide says "no API differences" but there are.
- Not clear how to migrate.
- Questionable API.
- Large changes to codebase (or we need to use a decorated HttpModule to maintain backward compatibility).

### Minimum reproduction code

Bump to NestJS 8 per migrate guide. See warning in IDE.
```ts
this.httpService.get(serviceUrl).toPromise();
```

### Expected behavior

`@nestjs/axios` should provide promise-based API.

**Comment by kamilmysliwiec** [comment]
> With toPromise() being deprecated, we are left with an Observable - this is quite a big API change!

According to RxJS migration guide, you should be able to replace `toPromise()` with the `lastValueFrom`.

Prev:

```typescript
this.httpService.get(serviceUrl).toPromise();
```

Current:

```typescript
lastValueFrom(this.httpService.get(serviceUrl))
```

**Comment by Izhaki** [comment]
Shouldn't the migration guide be updated?

And I can't help the feeling that this conclusion isn't great for developers using NestJS - we are required to change many lines of code and add imports directly from RxJS - just to get what seems to be the sensible API.

It also means this module diverges from the NestJS API, [which states](https://docs.nestjs.com/controllers#asynchronicity):

> The above code is fully valid. Furthermore, Nest route handlers are even more powerful by being able to return RxJS observable streams. Nest will automatically subscribe to the source underneath and take the last emitted value (once the stream is completed).

So this is the case for NestJS, but not `@nestjs/axios`.

**Comment by kamilmysliwiec** [comment]
> Shouldn't the migration guide be updated?

`toPromise()` is deprecated, not removed. Also, this change comes from the RxJS so you should navigate to the RxJS migration guide for more information on the breaking changes they've made.

> The above code is fully valid. Furthermore, Nest route handlers are even more powerful by being able to return RxJS observable streams. Nest will automatically subscribe to the source underneath and take the last emitted value (once the stream is completed).

This statement is still correct.

Please, use our [Discord](https://discord.gg/G7Qnnhy) channel (support) for such questions. We are using GitHub to track bugs, feature requests, and potential improvements.
